src/ChainlinkOracleImpl.sol

```diff
@@ -0,0 +1,308 @@
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+    /// set token override
+    function _setTokenOverride(SetTokenOverrideParams calldata params_) internal {
+        $_tokenOverrides[params_.token] = params_.tokenOverride;
+    }
+
+    /// set pair details
+    function _setPairDetails(SetPairDetailParams[] calldata params_) internal {
+        $_pairDetails._set(_convert, params_);
+    }
+
+    /// -----------------------------------------------------------------------
+    /// functions - private & internal - views
+    /// -----------------------------------------------------------------------
+
+    /// get amount for a quote
+    function _getQuoteAmount(QuoteParams calldata quoteParams_) internal view returns (uint256) {
+        ConvertedQuotePair memory cqp = quoteParams_.quotePair._convert(_convert);
+
+        // skip oracle if converted tokens are equal
+        if (cqp.cBase == cqp.cQuote) {
+            return quoteParams_.baseAmount;
+        }
+
+        SortedConvertedQuotePair memory scqp = cqp._sort();
+        PairDetail memory pd = $_pairDetails._get(scqp);
+        if (address(pd.feed)._isEmpty()) {
+            revert InvalidPair_FeedNotSet(quoteParams_.quotePair);
+        }
+        if (pd.staleAfter == 0) {
+            pd.staleAfter = $defaultStaleAfter;
+        }
+
+        (
+            , /* uint80 roundId, */
+            int256 answer,
+            , /* uint256 startedAt, */
+            uint256 updatedAt,
+            /* uint80 answeredInRound */
+        ) = pd.feed.latestRoundData();
+
+        if (updatedAt < block.timestamp - pd.staleAfter) {
+            revert StalePrice(pd.feed, updatedAt);
+        }
+        if (answer < 0) {
+            revert NegativePrice(pd.feed, answer);
+        }
+
+        uint256 quoteAmount;
+        uint8 answerDecimals = pd.feed.decimals();
+        bool pairInverted = (cqp.cBase != scqp.cToken0);
+        bool invertFeed = pd.inverted != pairInverted;
+        if (invertFeed) {
+            quoteAmount = uint256(quoteParams_.baseAmount) * 10 ** uint256(answerDecimals) / uint256(answer);
+        } else {
+            quoteAmount = uint256(quoteParams_.baseAmount) * uint256(answer) / 10 ** uint256(answerDecimals);
+        }
+        return _adjustQuoteDecimals(quoteAmount, quoteParams_);
+    }
+
+    /// adjust quoteAmount_ based on decimals of base & quote
+    function _adjustQuoteDecimals(uint256 quoteAmount_, QuoteParams calldata quoteParams_)
+        internal
+        view
+        returns (uint256)
+    {
+        uint8 baseDecimals = quoteParams_.quotePair.base._decimals();
+        uint8 quoteDecimals = quoteParams_.quotePair.quote._decimals();
+
+        int256 decimalAdjustment = int256(uint256(quoteDecimals)) - int256(uint256(baseDecimals));
+        if (decimalAdjustment > 0) {
+            return quoteAmount_ * 10 ** uint256(decimalAdjustment);
+        }
+        return quoteAmount_ / 10 ** uint256(-decimalAdjustment);
```

Comment:
just double checking this works ok if it's 0 (not sure if `-0` will cause an issue.